### PR TITLE
Fix viewness propagation

### LINKED ISSUE #191: np.allclose modifies ArrayView if it contains nan
This is weird, and probably not too common, but it suggests a bigger problem with ArrayView behavior:

```python
import scanpy as sc
import anndata as ad
from anndata.tests.helpers import assert_equal
pbmc = sc.datasets.pbmc68k_reduced()

import scanpy as sc
import numpy as np
# This dataset has nans in .varm["PCs"]
pbmc = sc.datasets.pbmc68k_reduced()
a = pbmc[:100].copy()
b = pbmc[:100]

# .copy() turns view to ndarray, so this is fine:
assert np.allclose(a.varm["PCs"], b.varm["PCs"].copy(), equal_nan=True)
# Next line triggers the mutation:
assert not np.allclose(a.varm["PCs"], b.varm["PCs"], equal_nan=True)
# Showing that the mutation happened:
assert not np.allclose(a.varm["PCs"], b.varm["PCs"].copy(), equal_nan=True)
```

It seems to me that the only way this could be happening is through `__setitem__` as making a dummy subclass of `np.ndarray` doesn't trigger the behavior. But why would `np.allclose` – also `np.isclose` – call `__setitem__`?

### LINKED ISSUE #239: indexing error on sliced X
```
import anndata
import numpy as np

adata = anndata.AnnData(np.random.random((10,10)))
adata = adata[:, [0,2,4]]
v = adata.X.var(axis=0)
v[np.isnan(v)] = 0
```

raises IndexError: boolean index did not match indexed array along dimension 0; dimension is 10 but corresponding boolean dimension is 3

### PR DESCRIPTION
**Pull request by flying-sheep**
Implementation mainly follows the example in https://numpy.org/devdocs/user/basics.subclassing.html#array-ufunc-for-ufuncs

@ivirshup: should we implement this for other classes?

Fixes #191, fixes #239

### CODE REVIEW

**Comment by flying-sheep** [comment]
Huh, weird, locally I get no errors at all.

**Comment by ivirshup** [comment]
I've updated the branch, there's a dependency issue. But I've just pinned it so it won't block CI.

File: anndata/_core/views.py [hunk 1]
```diff
@@ -103,6 +106,44 @@ def __array_finalize__(self, obj: np.ndarray | None):
...
+        results = super().__array_ufunc__(
+            ufunc, method, *convert_all(inputs), out=out, **kwargs
+        )
+        if results is NotImplemented:
+            return NotImplemented
+
+        if ufunc.nout == 1:
+            results = (results,)
+        results = tuple(
+            (np.asarray(result).view(np.ndarray) if output is None else output)
```
Comment (ivirshup) [inline anndata/_core/views.py:142]: Why do the type coercion of the result?

**Review by ivirshup — COMMENTED** [review]
> should we implement this for other classes?

Would this fix bugs in behavior with other classes? I'd lean toward no, if we don't have examples of that.

- [ ] It would probably also be good to run the scanpy test suite against this PR to check that it doesn't introduce other unintentional behavior.

> Fixes https://github.com/scverse/anndata/issues/191

- [x] Could you add tests for this?

File: anndata/tests/test_views.py [hunk 2]
```diff
@@ -545,12 +545,12 @@ def test_invalid_scalar_index(adata, index):
         _ = adata[index]
 
 
-@pytest.mark.parametrize("obs", [False, True])
+@pytest.mark.parametrize("axis", ["obs", "vars"])
 @pytest.mark.parametrize("index", [-100, -50, -1])
-def test_negative_scalar_index(adata, index: int, obs: bool):
-    pos_index = index + (adata.n_obs if obs else adata.n_vars)
+def test_negative_scalar_index(adata, axis, index):
+    pos_index = index + getattr(adata, f"n_{axis}")
```
Comment (ivirshup) [inline anndata/tests/test_views.py:551]: I'm not sure the changes here make the test better. Maybe lets revert?

File: anndata/tests/test_views.py [hunk 3]
```diff
@@ -565,6 +565,15 @@ def test_negative_scalar_index(adata, index: int, obs: bool):
     )
 
 
+def test_viewness_propagation():
```
Comment (ivirshup) [inline anndata/tests/test_views.py:568]:
```suggestion
def test_viewness_propagation():
    # https://github.com/scverse/anndata/issues/239
```

It's good to link the relevant issue for tests so we can go back for context.

File: anndata/tests/test_views.py (hunk 2 above)
Comment (flying-sheep) [reply anndata/tests/test_views.py:551]: I just need to stop doing aesthethic things like this in unrelated PR and do them independently

**Comment by flying-sheep** [comment]
> Could you add tests for this?

the test was the first commit.

**Comment by flying-sheep** [comment]
This is a bugfix, so 0.9.1 should be the correct milestone

**Comment by flying-sheep** [comment]
all done, merging

File: docs/release-notes/0.9.2.md [hunk 4]
```diff
@@ -1,4 +1,5 @@
 ### 0.9.2 {small}`the future`
 
 ```{rubric} Bugfix
+* Fix ufuncs of views like ``adata.X[:10].cov(axis=0) returning views {pr}`1043` {user}`flying-sheep`
```
Comment (ivirshup) [inline docs/release-notes/0.9.2.md:4]: Look at the other release notes for usage

File: anndata/_core/views.py [hunk 5]
```diff
@@ -103,6 +106,44 @@ def __array_finalize__(self, obj: np.ndarray | None):
...
+        results = super().__array_ufunc__(
+            ufunc, method, *convert_all(inputs), out=out, **kwargs
+        )
+        if results is NotImplemented:
+            return NotImplemented
+
+        if ufunc.nout == 1:
+            results = (results,)
+        results = tuple(
+            (np.asarray(result) if output is None else output)
```
[the remark this answers is not in the document]
Comment (flying-sheep) [reply anndata/_core/views.py:142]: I did

**Comment by flying-sheep** [comment]
everything was resolved

File: docs/release-notes/0.9.2.md (hunk 4 above)
Comment (flying-sheep) [reply docs/release-notes/0.9.2.md:4]: ???

Comment (ivirshup) [reply docs/release-notes/0.9.2.md:4]: Text doesn't go inside the ticks:

https://raw.githubusercontent.com/scverse/anndata/main/docs/release-notes/0.8.1.md

File: anndata/_core/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/_core/views.py:142]: `np.asarray`?

**Comment by ivirshup** [comment]
But resolved by you, not by the reviewer. Clearly we understood the coercion change differently, and there was a bug in the addition of the release notes.

**Pull request merged**